Audio to web

luisesn opened the pull request:
This is not intended to merge to main.

Its a work in progress so the app can receive and play audio sent from the radio.

For now it works with an standalone example (soundWs/index.html) from your local computer, and I have started to implement it in the vue app but for now it only works locally, not from the device (but as I dont have any idea of how to program in Vue... it'll have to do for today)

kamilsss655 commented:
That's an interesting concept. Will take a look!

kamilsss655 commented:
Can you provide some higher level overview on how this works?
Is it just sending audio samples over websocket or is there some audio encoding happening in between?
I see some sum of squares stuff etc

luisesn commented:
At the esp level i've just copied your recording function so its sends the data over ws.
I have created a version of your ws send so it sends binary data too (but there is some error somewhere because sometimes the packets are corrupt and ws connection closes, didn't have time but I suspect it has to do with the way the buffers are treated.
)
I have modified the audio data block size, audio rate and converted data from 16 bit to 8 bit because I dont need the quality for now and preferred the speed and less data (256kbps to 64 kbps reduction aprox.) (maybe if we find some way of compressing the data).


At the web side its only a websocket that receives the data and buffers it so it can be played (with some "apaños" because it wasn't detecting the binary data right).

The sum of squares its only to show the volume in a progress bar, the only needed operation its converting from int values to -1.0/1.0 ones so audiocontext can understand (the /128 in the current version).

kamilsss655 commented:
Thanks for the explanation. Makes a lot of sense.

I haven't tried it yet, but will do soon!

kamilsss655 commented:
I tested it and it looks good so far. Hope you can make it work from the device.

The audio was glitching for me, but I guess that is part of the WIP.

Really like the idea and hope you can make it work. It would make a fine addition to the feature set of ESPRI.

Good luck!

luisesn commented:
The glitching is caused by the websocket closing I'm hoping to correct it... 
I'll keep working on it

kamilsss655 commented:
This might be irrelevant, but I noticed a lot of log messages being output over serial during audio listening. Perhaps disabling these might help?

luisesn commented:
it should not affect as the sound is processed in background in another thread

the sttutering is mostly caused by the websocket closing (maybe 1% of the ocassions is network speed/lag/cuts, TCP its not the best transport for this but... mostly the closing)
